src/1-theory/3-input-and-output.adoc

```diff
@@ -0,0 +1,593 @@
+encoding, for an image means resizing, etc.)
+* 400 - Bad Request (should really be for invalid syntax, but some folks
+use for validation)
+* 401 - Unauthorized (no current user and there should be)
+* 403 - The current user is forbidden from accessing this data
+* 404 - That URL is not a valid route, or the item resource does not
+exist
+* 405 - Method Not Allowed (your framework will probably do this for
+you)
+* 409 - Conflict (Maybe somebody else just changed some of this data, or
+status cannot change from e.g: "published" to "draft")
+* 410 - Data has been deleted, deactivated, suspended, etc.
+* 415 - The request had a `Content-Type` which the server does not know
+how to handle
+* 429 - Rate Limited, which means take a breather, sleep a bit, try
+again
+* 500 - Something unexpected happened, and it is the APIs fault
+* 503 - API is not here right now, please try again later
+
+You might spot others popping up from time to time, so check on
+http://http.cat/[http.cats] (or
+http://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml[iana.org]
+for a more formal list) when you see one that's not familiar.
+
+== Mime Types
+
+HTTP APIs can work with all sorts of data. Whilst SOAP may have been restricted
+to XML, REST and GraphQL can work with any response types. gRPC is kinda stuck
+with Protobuf.
+
+An API can support almost unlimited options, but of course building support for
+every content type ever would be a rather laborious job. There are a few we can
+cut out early on.
+
+=== Ditch Form Data
+
+"Form Data" uses the `application/x-www-form-urlencoded` mime type, and mostly
+only seems to be used by PHP developers. Luckily most other folks ignore this
+wholeheartedly.
+
+.Form data doesn't really have data types, just awkward strings.
+[source]
+----
+foo=something&bar=1&baz=0
+----
+
+Another issue with form data is similar to how XML suffers a lack of obvious
+data types. For example, to handle a boolean a client has to send `1` or `0`,
+because `bar=true` would be `string("true")` on the server-side.
+
+Data types are important, so let's not just throw them out the window for the
+sake of "easy access to our data", especially as most web application frameworks
+have something like `$request->body->foo` to easily get to the foo property.
+
+WARNING: Rails is awful at this. If you have a `?foo=a` query string parameter,
+and you also send `{ "foo": "b" }` in the HTTP body, then `params[:foo]` will be
+set to `"b"` as the latter overrides the former. Code that you build, including
+any generic frameworks/tooling that you release, should avoid conflating query
+strings and body properties at all costs. They're different things and this
+nonsense causes confusion, especially when you realize that `params[:action]`
+means "controller method name" and actually overrides whatever is in `?action=`
+_without_ anything being in the body...
+
+.Instead of form data, use a nice JSON object.
```

Comment:
Maybe it's worth including an example of this request using form data before you give the JSON example? If anything the length of the URL (and lol URls have max length which sometimes bites your form data in the ass!) should put people off enough?

Reply:
LOL I really need to read on before commenting! In any case for some reason the bad example before the good example could work better, but that might just be me, so feel free to ignore. 😇 

The point about URLs having a limit could be good to mention. It bit us when we worked with Magento's API and they wanted a ton of stuff posted as a form, and the descriptions of products can get quite long... took me a while to work out the browser was just cutting it off after ~2000 chars!  (https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers)